scripts/bqeth_subscription.py

```diff
@@ -0,0 +1,266 @@
+        return func(context, *args, **kwargs)
+
+    return wrapper
+
+
+@click.group()
+@click.option("--domain", default="lynx", help="NuCypher domain.")
+@click.pass_context
+def cli(ctx, domain):
+    """BqETH Subscription CLI"""
+    load_dotenv(override=True)
+    provider_url = os.environ.get("PROVIDER_URL")
+    private_key = os.environ.get("PRIVATE_KEY")
+    ctx.ensure_object(dict)
+    ctx.obj["constants"] = build_constants(domain, provider_url, private_key)
+
+
+@cli.command()
+@click.option("--encryptor-slots", default=2, help="Number of encryptor slots to pay for.")
+@setup_context
+def pay_subscription(context, encryptor_slots):
+    """Pay for a new subscription period and initial encryptor slots."""
+    base_fees = context.subscription_contract.functions.baseFees(0).call()
+    encryptor_fees = context.subscription_contract.functions.encryptorFees(
+        encryptor_slots, context.subscription_contract.functions.subscriptionPeriodDuration().call()
+    ).call()
+
+    click.echo(
+        f"Approving transfer of {base_fees + encryptor_fees} ERC20 token for subscription contract."
+    )
+    tx_hash = context.erc20_contract.functions.approve(
+        context.constants["SUBSCRIPTION_CONTRACT_ADDRESS"], base_fees + encryptor_fees
+    ).transact({"from": context.account.address})
+    click.echo(f"Transaction hash: {tx_hash.hex()}")
+
+    click.echo(f"Paying for a new subscription period with {encryptor_slots} encryptor slots.")
+    tx_hash = context.subscription_contract.functions.payForSubscription(encryptor_slots).transact(
+        {"from": context.account.address}
+    )
+    click.echo(f"Transaction hash: {tx_hash.hex()}")
+
+
+@cli.command()
+@click.option("--extra-slots", default=1, help="Number of additional encryptor slots to pay for.")
+@setup_context
+def pay_slots(context, extra_slots):
+    """Pay for additional encryptor slots."""
+    click.echo(f"Paying for {extra_slots} new encryptor slots.")
+    tx_hash = context.subscription_contract.functions.payForEncryptorSlots(extra_slots).transact(
+        {"from": context.account.address}
+    )
+    click.echo(f"Transaction hash: {tx_hash.hex()}")
+
+
+@cli.command()
+@click.option("--num-nodes", default=2, help="Number of nodes to use for the ritual.")
+@setup_context
+def initiate_ritual(context, num_nodes):
+    """Initiate a ritual."""
+    nodes = list(sorted(
+        [
+            u["checksum_address"]
+            for u in requests.get(
+                f"{context.constants['PORTER_ENDPOINT']}/get_ursulas?quantity={num_nodes}"
+            ).json()["result"]["ursulas"]
+        ]
+    ))
+    start_of_subscription = context.subscription_contract.functions.startOfSubscription().call()
+    duration = (
+        context.subscription_contract.functions.subscriptionPeriodDuration().call()
+        + context.subscription_contract.functions.yellowPeriodDuration().call()
+        + context.subscription_contract.functions.redPeriodDuration().call()
+    )
```

Comment:
Ah - This is very interesting functionality!  Just making sure I understand correctly it looks like you are correlating a ritual's duration with the specific temporal rules of a subscription?

Reply:
@theref Can you make a commit into PR #297 to include these calculations?

Reply:
#297 is merged now but it will still be good to include these changes in `initiate_ritual.py`  @theref Do you want to take care of that?